### PR TITLE
Add `StructuredCloneable` type

### LINKED ISSUE #855: Type for serializable objects?
Would it make sense to create a type for [serializable objects](https://developer.mozilla.org/en-US/docs/Glossary/Serializable_object)? 

Similar to Jsonify et al., this would match values that could losslessly be deep-copied via `structuredClone` or `postMessage`.

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/855">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/855/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by thecodewarrior**
This adds two types for values losslessly cloneable using `structuredClone`.

Notes:
- The web and non-web types are separate so people who aren't writing browser code don't pull in web types
- Custom error classes do match, despite the fact that `structuredClone` isn't able to losslessly clone them (they're cloned as base `Error` instances). After doing some research I couldn't figure out a way to include the root Error class but exclude subclasses, so I just accepted the limitation and documented the discrepancy.

Fixes #855

### CODE REVIEW

File: source/structured-cloneable.d.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	| Date
+	| Error
+	| RegExp
+	| TypedArray;
+
+type StructuredCloneableCollection =
+	| StructuredCloneable[]
+	| {[k: string]: StructuredCloneable; [k: number]: StructuredCloneable}
+	| Map<StructuredCloneable, StructuredCloneable>
+	| Set<StructuredCloneable>;
```
Comment (sindresorhus) [inline source/structured-cloneable.d.ts:26]: I would use the read-only type versions here so it can support read-only collections too.

**Comment by Emiyaaaaa** [comment]
I think this PR needs to wait for #898 to merge.

**Comment by sindresorhus** [comment]
Tests are failing now. Not sure how we can handle DOM types conditionally.

**Comment by sindresorhus** [comment]
Relevant: https://github.com/microsoft/TypeScript/issues/31894

**Comment by sindresorhus** [comment]
I think we could do it by adding a `dom` sub-export in package.json: https://nodejs.org/api/packages.html#subpath-exports

**Comment by thecodewarrior** [comment]
Adding the sub-export would probably involve changes to the testing/build infrastructure (running twice for non-dom/dom builds), which I'm not familiar with, but if you want to implement that I'd be happy to add back the web type. If we're doing a sub-export I think it wouldn't even need to have a separate name, just `import {StructuredCloneable} from 'type-fest/dom'` instead of `import {StructuredCloneable} from 'type-fest'`

**Comment by sindresorhus** [comment]
Thanks :)

**Pull request merged**